### ISSUE TITLE
Add support for new HMAC-256 signature calculation

### ISSUE THREAD

**Issue by prognostikos** [wvanbergen/adyen#118]
As noted in the HPP changelog[1], Adyen has introduced a new recommended way to calculate the merchantSig for HPP requests.

I will be submitting a PR - this is just a placeholder until then.

@wvanbergen I would like your input on this, though, if you have time - I imagine making the signature method for a skin configurable, defaulting to the existing `Adyen::Util.hmac_base64` method. If you'd prefer to default to the new way let me know.

[1]:https://docs.adyen.com/display/TD/HPP+manual+changelog

**Comment by wvanbergen** [comment]
Is there any reason to keep the old signing mechanism around?

**Comment by prognostikos** [comment]
Hm, well, Adyen still allows you to create new skins with the deprecated
signing method, but I have no problem saying master only supports the new
method.

Matt Rohrer
matt@prognostikos.com


On Thu, Sep 3, 2015 at 11:32 AM, Willem van Bergen <notifications@github.com

**Comment by wvanbergen** [comment]
Ah: so it's not either/or, you have to pick one for your skin?

**Comment by prognostikos** [comment]
Yeah, this is what you see after clicking "new skin" 
![image](https://cloud.githubusercontent.com/assets/515254/9660166/85debabc-5254-11e5-875d-446faa769c9f.png)

**Comment by prognostikos** [comment]
And actually, I imagine that I can only chose the old one because I already have skins using it in production. I imagine new users don't even get a choice.

**Comment by wvanbergen** [comment]
given that this is a major release, I have no problem making the new method default, and keeping the old one around as an option. :+1:

**Comment by prognostikos** [comment]
Now that I've looked at the Adyen::Form code and the different signing methods (payment, billing_address, shopper_signature, etc.), I wonder if it might make sense to only support the new signing method with the new Adyen:HPP class and leave the old methods in the (deprecated) Adyen::Form module.

The new way is much more straightforward and will remove a lot of code, as it accepts an arbitrary hash of parameters and just sorts and concatenates all of the keys and values for signing, as opposed to the old way where specific values needed to be concatenated in a given order for each of the signature types.

**Comment by wvanbergen** [comment]
sounds good to me :+1:

**Comment by thailehuy** [comment]
May I ask when can this be merged? I checked the code, the signature calculation is correct, but the redirect verify function is not correct yet

**Comment by wvanbergen** [comment]
@thailehuy Maybe you can help getting the verify signature method to work? @prognostikos: any progress on you `Adyen::HPP` work?

**Comment by prognostikos** [comment]
@wvanbergen unfortunately #109 has been bumped by other things so it will be a while before I can devote any time to it. @thailehuy we had discussed only supporting the new signing method as part of the refactor to `Aden::HPP` in #109 so you're welcome to contribute there or here - whichever makes most sense for your app.

I will note that the new system seems to be only partially supported by Adyen's debugging tools in their CA which is part of the reason why I've delayed on this and just kept creating skins with the old signing method. Even though they're calling it deprecated they have not yet built up the same level of tooling for the new method.

**Comment by thailehuy** [comment]
@prognostikos I just found out that the return signing string is correct (the one that is joined with `:`), but the form signing string is still this one
```
def calculate_signature_string(parameters)
      merchant_sig_string = ""
      merchant_sig_string << parameters[:payment_amount].to_s       << parameters[:currency_code].to_s        <<
                             parameters[:ship_before_date].to_s     << parameters[:merchant_reference].to_s   <<
                             parameters[:skin_code].to_s            << parameters[:merchant_account].to_s     <<
                             parameters[:session_validity].to_s     << parameters[:shopper_email].to_s        <<
                             parameters[:shopper_reference].to_s    << parameters[:recurring_contract].to_s   <<
                             parameters[:allowed_methods].to_s      << parameters[:blocked_methods].to_s      <<
                             parameters[:shopper_statement].to_s    << parameters[:merchant_return_data].to_s <<
                             parameters[:billing_address_type].to_s << parameters[:offset].to_s
    end
```

So the code is all good, sorry for the trouble

**Comment by arnoutdemooij** [comment]
@thailehuy In addition to your remark about slicing Adyen's return parameters when using Rails:
By default, Rails also includes extra fields (named 'authenticity_token' and 'utf8') in forms constructed with `form_tag`. Adyen apparently includes these fields in it's own signature calculation, which results in a different signature on their side and the request being rejected. To prevent this, use `form_tag(<url>, authenticity_token: false, enforce_utf8: false)` (NB the param `:enforce_utf8`, is not available in older versions of Rails).
With this precaution, SHA-256 based form signing (with the colon-delimited signing string) can be used successfully.

**Comment by thailehuy** [comment]
@arnoutdemooij I didn't encounter that `authenticity_token` issue, but I guess precaution can be taken

One more thing, Adyen may return the encoded `+` character in the signature, which will be translated to an empty space by the web server before the params reach the signature comparison code

So it might be a good idea to gsub space for `+` in the params `merchantSig`

**Comment by wvanbergen** [comment]
maybe we should add a `form_tag` helper that works around this issues?

**Comment by arnoutdemooij** [comment]
Agree.
I also had to remove the 'name' attribute from my form's submit button (input type='submit') because otherwise it's value would be included in Adyen's signature calculation...

**Comment by aponsin** [comment]
Sorry to get in the middle of the discussion here. I did not know where else to ask my question. I realize that you are still working on bringing support to the new skin type (with SHA-256 encryption). Adyen has basically told us that we need to move to the new skin type, because the other (older) version of the skin does not allow them to configure our contract properly. So we've been looking at implementing the new skin. So far I have been trying to test the HPP branch to get that to work in a testing environment. But no luck so far. I was under the impression that the testing would work, even-tough the PR is not ready for merging. Am I missing something ? Could you give me some pointers as to how to use the new skin type with the HPP branch ? Much appreciated.

**Comment by arnoutdemooij** [comment]
Hi Alexandre,

Below follows an overview of how I'm using the HPP branch. At what point do you get stuck exactly? Any errors?

---

I'm using the HPP branch of my fork of the Adyen gem in a Rails 3 project (v3.2.22) with Ruby 2.0.0.

Gemfile:

```ruby
gem 'adyen', git: 'https://github.com/arnoutdemooij/adyen.git', branch: 'hpp'
````
Be sure Gemfile.lock specifies the latest revision for the gem (currently 4e630c9708b0fb014077d0b63b76bda1a74d4062)

On Adyen's Customer Area (https://ca-test.adyen.com) I have created a skin with an automatically generated SHA-256 HMAC key for the test platform.

I have added a file config/initializers/adyen.rb to my project:

```ruby
Adyen.configuration.default_form_params[:merchant_account] = ENV['ADYEN_MERCHANT_ACCOUNT']
Adyen.configuration.default_skin = :default
Adyen.configuration.register_form_skin(:default, ENV['ADYEN_SKIN_CODE'], ENV['ADYEN_SKIN_SECRET'])
```

I have a view partial to render the form that will be POST-ed to Adyen which contains (HAML):

```ruby
- hpp_request = Adyen::HPP::Request.new(payment_data, environment: 'test')

%form{ 'accept-charset' => 'UTF-8', 'action' => hpp_request.url, 'method' => :post }
  = hpp_request.hidden_fields
  = submit_tag(t('.pay'), name: nil)
```
I use a form tag instead of the form_tag helper, to avoid Rails adding additional hidden fields to the form, such as 'authentication_token'. Also, I make sure that the submit button has no name attribute set.

`payment_data` is a hash, something like:

```ruby
    {
      currency_code: 'EUR',
      ship_before_date: Time.zone.now.to_date + 1.day,
      session_validity: Time.zone.now + 30.minutes,
      recurring: false,
      merchant_reference: <string>,
      payment_amount: <integer>,
      shopper_locale: 'nl_NL'
    }
```
The skin code is not included here, because it is added based on the skin I specified in the configuration. To be explicit I also could have written in my view:

```ruby
- hpp_request = Adyen::HPP::Request.new(payment_data, environment: 'test', skin: :default)
```

Or instead I could have included a skin_code param in the payment data and supplied the shared secret directly, like:

```ruby
- hpp_request = Adyen::HPP::Request.new(payment_data_including_skin_code, environment: 'test', shared_secret: <string>)
```

The form that is generated is POST-ed to https://test.adyen.com/hpp/select.shtml successfully, which means the calculated (SHA-256) merchant signature is correct. When I complete the (fake) payment, it becomes visible in the Payments section of the Adyen Customer Area (CA).

The customer is redirected by Adyen to the Result URL specified in the skin details in the CA. This redirect is handled by the `result` action of my payments controller. There, the response params are checked with:

```ruby
class Public::PaymentsController < ApplicationController
  before_filter :check_signature, only: :result

  # ...

  private

  def response_params
    params.slice(*%w(authResult merchantReference merchantReturnData paymentMethod pspReference shopperLocale skinCode merchantSig))
  end

  def check_signature
    raise "Invalid response!" unless Adyen::HPP::Response.new(response_params).has_valid_signature?
  end
end
```
Note that the params are 'sliced' to remove any parameters added by Rails, such as the controller and action name.

Adyen also sends a server-to-server notification of the payment result. This is handled by the `postback` action of the same payments controller. The notification data in the params is not signed and the format is explained on https://docs.adyen.com/display/TD/Notification+examples.

**Comment by stulentsev** [comment]
Thanks, @arnoutdemooij, it helped a lot.

**Comment by wvanbergen** [comment]
It's crunch time for me at work, so unfortunately I won't have a lot of time to look at this during the next couple of weeks. Sorry!

**Comment by hasanen** [comment]
What's the status of this?

**Comment by LaurentVB** [comment]
Hello, 
Adyen just issued a warning that the old merchant signature encoding using SHA-1 will not be supported starting July 1st 2016. We all need to upgrade...
Any progress on this issue?

**Comment by mvz** [comment]
Anyone upgrading to SHA-256 please test the `hpp` branch and report your results.

**Comment by theindra** [comment]
I was trying to generate the redirect_url for adyen hpp and it worked with HMAC-256. 

But i can't use the gem with jruby 1.7 like the old version pre 2.0.0 version. It's always a syntax error. Could you add support for old ruby syntax?

**Comment by mvz** [comment]
@theindra Is there any particular reason why you're not upgrading to the JRuby 9.0 series?

**Comment by theindra** [comment]
Because our whole ecosystem runs on jruby 1.7 and we don't have enough resources to test all features and services if they are working. The deadline would be 1. July, because then adyen will stop the old hmac signature

**Comment by mvz** [comment]
@theindra Ok, I'll take a look this week.

**Comment by wvanbergen** [comment]
@theindra I don't think we should go out of our way to keep supporting old Ruby syntax. 

The problematic feature is using keyword arguments, which are supported from Ruby 2.0 and up, which was released a long time ago already. Keyword arguments are a very useful language feature that will make for more robust code. E.g., if you accidentally make a typo in a keyword argument name, it will blow up immediately, while with option hashes you end up with a hard to debug problem. Also, the code is more self-documenting when using keyword arguments.

So using this language is a good thing, and important for the long-term health of a project like this, and I don't think we should revert our usage of it. That means you'll have to do some work on your end. Whether that means upgrading to a newer version of JRuby, or maintaining your own fork of this gem (use https://github.com/wvanbergen/adyen/pull/147 as a starting point), is up to you. 

But you can't expect us as volunteers to hold back this project, or put in the time to find workarounds to support old unsupported Ruby versions, because you are not able or willing to put in the time yourself.

**Comment by mvz** [comment]
> maintaining your own fork of this gem (use #147 as a starting point)

I also think this is the best solution, @theindra. To do this, you should create your own fork of this repository and maintain the necessary changes there.

**Comment by SebAshton** [comment]
Hi @mvz Is there any reason the `Adyen::Form` is still using the old `Adyen::Util.hmac_base64` to calculate the merchant signature rather than the new `Adyen::Signature`?

Is there a new recommended way to handle HPPs?

**Comment by mvz** [comment]
@SebAshton Adyen::Signature is a very recent addition. Adyen::Form probably won't be updated to use it, since it only supports the old signature that Adyen is phasing out.

**Comment by SebAshton** [comment]
@mvz any update when the `hpp` branch will be merged into master and a version released?

**Comment by mvz** [comment]
@SebAshton I can't release, so it's up to @wvanbergen.

**Comment by wvanbergen** [comment]
Released version 2.1.0. @mvz if you send me your rubygems email, I can add you as an owner as well.

**Comment by wvanbergen** [comment]
Going to close this issue; if there's any specifics around signature calculation that still need improvement or fixing, please open another issue.

**Comment by SebAshton** [comment]
@wvanbergen there is code in the `hpp` branch that aren't in `master` did the `2.1.0` release come from that (the `hpp`) branch?

**Comment by wvanbergen** [comment]
No, I always release from the master branch. I was assuming all necessary changes were merged.

**Comment by SebAshton** [comment]
@wvanbergen who can merge it in? Is that yourself or @mvz?

**Comment by wvanbergen** [comment]
@mvz should be able to merge it. I don't really have enough context to know whether it is complete or not.

**Comment by mvz** [comment]
I'll have a final check tomorrow and merge it.

**Comment by SebAshton** [comment]
@mvz how is this looking?

**Comment by mvz** [comment]
@SebAshton I just updated the branch with master, and there are still some tests failing. Looking into that now.

**Comment by mvz** [comment]
@SebAshton ... and it's merged.

@wvanbergen did something change with your integration test credentials? Some of the integration specs are failing on master.

**Comment by wvanbergen** [comment]
I asked Adyen to enable payouts (#148) on the test account. Maybe that broke something? I may be able to look into it later today or tomorrow.

**Comment by mvz** [comment]
Ok.

**Comment by reneweteling** [comment]
Thanks guys, keep us posted 👍

**Comment by wvanbergen** [comment]
I got a passing test run again on master. It appears like the functional and integration tests can be somewhat flaky.

**Comment by wvanbergen** [comment]
I released version 2.2.0 including `Adyen::HPP`, which supports the new signing mechanism.
